### ISSUE TITLE
Feature Request: Scene Detection

### ISSUE THREAD

**Issue by DarkStarSword** [bo3b/3Dmigoto#33]
Want a way to detect scenes based on active shaders or textures to set custom convergence and ini params.

We can already partially do this as a consequence of some existing features, but it's too limited to be generally useful. e.g. we can have a shader set an IniParam and clear it later on a Present call, but that only gives us scene detection in part of the frame, and doesn't give us the ability to change the convergence.

Want to reuse the code in Override.cpp for this as much as possible
  - Allow the use of the existing transition code
  - Share the global undo with key bindings
  - An override should activate as soon as the shader/texture is encountered (or at the start of the next frame?)
  - Override should deactivate on the present call if an entire frame (or more?) has passed without it being triggered again.

We want to be able to trigger scene detection from both [ShaderOverride] and [TextureOverride] sections. It would be a performance nightmare to check every single texture slot of every shader type on every draw call. I therefore intend to require something in the [ShaderOverride] section to enable texture scene detection from a specific slot, in much the same way we currently do for texture filtering.

Ideally, the syntax would be the same as in the [Key] sections, but there are some problems:
- [ShaderOverride] sections already have x,y,z,w, etc for texture filtering, partner filtering, etc. that remain sticky
- [ShaderOverride] sections already have "convergence" and "separation" settings that affect a single draw call
- We could potentially do something to reuse these existing settings (if we have some way to specify a different duration), but some of the settings in [ShaderOverride] are now being processed as though they are instructions in sequential order, however the existing syntax in the [Key] sections don't fit with this as they have modifiers (like transition) on separate lines. We could make this work, but it would be an exception to how everything else works in these sections.

The best way I can see to do this, is something like this:

    [ShaderOverrideSceneDetect]
    preset = PresetSceneFoo
    
    [PresetSceneFoo]
    x = 1.0
    convergence = 0.5
    transition = 300
    transition_type = cosine
    release_transition = 100
    release_transition_type = cosine

In this way, we are able to use the same syntax as the [Key] sections with no ambiguity. preset could potentially also refer to existing [Key] sections. The downside is we now have x,y,z,w,convergence & separation in both the ShaderOverride and the Preset section it refers to, that may appear similar at first glance, but are subtly different. Good documentation on the feature will be the key here.

Another problem to consider (that exists in pure key bindings as well, but until now we could largely get away with ignoring) is what happens if two overrides are activated simultaneously that both change a setting to different values. e.g. for convergence
- The global undo area is there to prevent this situation from permanently changing the convergence - once all overrides are deactivated it is guaranteed to return to the value it was before any overrides were activated.
- If a single override is activated then deactivated there is no problem.
- If all overrides set the same value there is no problem.
- activate A, activate B, deactivate B, deactivate A: Current behaviour is mostly sensible, but what if B shouldn't override A?
- activate A, activate B, deactivate A, deactivate B: Current behaviour is not ideal - after A is deactivated convergence will be set for A, but probably should be set for B.

In order to improve this I have two ideas:
- Maintain a list of active overrides. When an override is disabled, remove it from the list and use whatever value the final entry in the remaining list is.
- Assign a priority to overrides. This could allow an aiming override to always take priority over a scene detection (or vice versa).

**Comment by DarkStarSword** [comment]
Will have to pay special attention to Crysis 3. Frame analysis often turns up empty in that game, which makes me suspect that it has excess present calls (perhaps due to the Origin overlay) and may be throwing out our idea of when a frame starts and ends.

**Comment by DarkStarSword** [comment]
[Preset*] and [Present] sections are only one letter apart and could easily be mistaken. Maybe better to use something else... Maybe [Override*], or is that too confusing since we already have [ShaderOverride] and [TextureOverride]?

**Comment by DarkStarSword** [comment]
type parameter from [Key] sections doesn't make sense for this. type=hold is kind of what we are doing, but rather it's "hold so long as the trigger fires at least once per frame". type=cycle and type=toggle definitely meaningless. Any need for type=activate so that the preset is never deactivated?

delay and release_delay could still make sense, but these won't automatically work as they use code in input.cpp, not override.cpp. Might be worth moving this code.

Might also be worth being able to specify a delay/release delay in frames instead of milliseconds, but that would mostly be for games that call present() multiple times in a frame, and is probably therefore better as a global setting to indicate how many presents must have passed without seeing the trigger before we can deactivate the override.

**Comment by DarkStarSword** [comment]
Another thing that might be nice, is to remember if the user changed the convergence/separation while an override was active, and the next time it is activated restore the same value.

Taking this further, we could write the value back to the d3dx.ini or maybe a d3dx-user.ini like Helix Mod does on F7

**Comment by DarkStarSword** [comment]
It looks like @llyzs has added support for this. I haven't personally reviewed or tested it yet, but can we either close this issue, or update it if there is anything left to do?

**Comment by bo3b** [comment]
Pretty sure this is working, and he is using it in Nier Automata.  

He is using Presets that is similar to the HelixMod approach.  And also allows the transition effects because it uses the command processing.

Not sure this meets all of our your requirements for the feature.  In particular, I don't think it has an F7 type save to change options in the d3dx.ini file.

Your choice on whether you'd rather close this or leave it open for later review.

**Comment by DarkStarSword** [comment]
I'll keep it open for now at least until I've played around with it a little.

**Comment by llyzski** [comment]
@DarkStarSword 

Sorry I did not notice you had opened the exact same issue here because I had just read the title and not knowing it was talking the exact same feature request... or I should have post my discussion here. :)

**Comment by DarkStarSword** [comment]
I've moved the preset activation over to the command list instead of ProcessShaderOverrides() because I need to use it based on textures, not just shaders (and to benefit from any future enhancements to the command list, such as conditional logic): ee5bad6f41ab7779e034af723889c5959b864e93

This does mean it will no longer work in conjunction with the index buffer filter - is that going to be a problem @llyzski? (the index buffer filter should really also be moved to the command list by extending the existing texture filtering feature to accept 'ib', as well as 'vb0', 'oD', etc.)

**Comment by DarkStarSword** [comment]
It seems that IniParams are not correctly restored when the preset is released...

**Comment by DarkStarSword** [comment]
The way the code is at the moment only one preset can be active at a time, and it will always be the last preset triggered during a frame. This doesn't seem ideal to me, since it precludes the possibility of having presets that affect independent variables.

**Comment by DarkStarSword** [comment]
The above limitation is now fixed by:
https://github.com/bo3b/3Dmigoto/commit/07d2211c500ef440114aafa043f3e2e3873307a1

If this is insufficient for anyone's needs, we might want to start thinking about improving the undo area and/or assigning priorities to presets to resolve conflicts when multiple presets are trying to set the same value.

**Comment by llyzski** [comment]
@DarkStarSword 

I am not actually using index buffer filter so this is not a problem for me, and as you said we can move it also to command list if someone else needs it. Thanks for the improvements.

**Comment by DarkStarSword** [comment]
Ok, good to know (and I'm not aware of anyone else using the index buffer filter either) :)

**Comment by DarkStarSword** [comment]
I've renamed this issue to reflect that the only remaining work left here is to improve how we handle conflicting overrides that have been activated simultaneously. This is pretty low priority, but would be nice to have.
